pkg/registry/contractRegistry.go

```diff
@@ -0,0 +1,215 @@
+	changedNodeNotifiersMutex sync.RWMutex
+}
+
+func NewSmartContractRegistry(
+	ethclient bind.ContractCaller,
+	logger *zap.Logger,
+	options config.ContractsOptions,
+) (*SmartContractRegistry, error) {
+	contract, err := abis.NewNodesCaller(
+		common.HexToAddress(options.NodesContractAddress),
+		ethclient,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &SmartContractRegistry{
+		contract:             contract,
+		refreshInterval:      options.RefreshInterval,
+		logger:               logger.Named("smartContractRegistry"),
+		newNodesNotifier:     newNotifier[[]Node](),
+		nodes:                make(map[uint16]Node),
+		changedNodeNotifiers: make(map[uint16]*notifier[Node]),
+	}, nil
+}
+
+/*
+*
+Loads the initial state from the contract and starts a background refresh loop.
+
+To stop refreshing callers should cancel the context
+*
+*/
+func (s *SmartContractRegistry) Start(ctx context.Context) error {
+	s.ctx = ctx
+	// If we can't load the data at least once, fail to start the service
+	if err := s.refreshData(); err != nil {
+		return err
+	}
+
+	go s.refreshLoop()
+
+	return nil
+}
+
+func (s *SmartContractRegistry) OnNewNodes() (<-chan []Node, CancelSubscription) {
+	return s.newNodesNotifier.register()
+}
+
+func (s *SmartContractRegistry) OnChangedNode(
+	nodeId uint16,
+) (<-chan Node, CancelSubscription) {
+	s.changedNodeNotifiersMutex.Lock()
+	defer s.changedNodeNotifiersMutex.Unlock()
+
+	notifier, ok := s.changedNodeNotifiers[nodeId]
+	if !ok {
+		notifier = newNotifier[Node]()
+		s.changedNodeNotifiers[nodeId] = notifier
+	}
+	return notifier.register()
+}
+
+func (s *SmartContractRegistry) GetNodes() ([]Node, error) {
+	s.nodesMutex.RLock()
+	defer s.nodesMutex.RUnlock()
+
+	nodes := make([]Node, len(s.nodes))
+	for idx, node := range s.nodes {
+		nodes[idx] = node
+	}
+	return nodes, nil
+}
+
+func (s *SmartContractRegistry) refreshLoop() {
+	ticker := time.NewTicker(s.refreshInterval)
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-ticker.C:
+			if err := s.refreshData(); err != nil {
+				s.logger.Error("Failed to refresh data", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (s *SmartContractRegistry) refreshData() error {
+	fromContract, err := s.loadFromContract()
+	if err != nil {
+		return err
+	}
+
+	newNodes := []Node{}
+	for _, node := range fromContract {
+		existingValue, ok := s.nodes[node.NodeId]
+		if !ok {
+			// New node found
+			newNodes = append(newNodes, node)
+		} else if !node.Equals(existingValue) {
+			s.processChangedNode(node)
+		}
+	}
+
+	if len(newNodes) > 0 {
+		s.processNewNodes(newNodes)
+	}
+
+	return nil
+}
+
+func (s *SmartContractRegistry) processNewNodes(nodes []Node) {
+	s.logger.Info("processing new nodes", zap.Int("count", len(nodes)), zap.Any("nodes", nodes))
+	s.newNodesNotifier.trigger(nodes)
+
+	s.nodesMutex.Lock()
+	defer s.nodesMutex.Unlock()
+	for _, node := range nodes {
+		s.nodes[node.NodeId] = node
+	}
```

Comment:
Can you walk through what happens in the scenario when a listener is notified of a new node, but the registry hasn't written that node into its map yet? (e.g. the listener calls `GetNodes()` first)

Reply:
I wouldn't expect the listener to need to do that (they already have all the new nodes). But maybe I should move the trigger call to be after just to be safe.